### PR TITLE
IGNITE-16801 added error handling during rebalance in case of errors from JRaft.

### PR DESCRIPTION
**Pull request by alievmirza**

### CODE REVIEW

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 1]
```diff
@@ -431,7 +431,10 @@ public Entry get(byte[] key, long rev) {
 
         try {
             return doGet(key, rev, true);
-        } finally {
+        } catch (NullPointerException e) {
```
Comment (sanpwc) [inline modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java:434]: Why we need this?

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 2]
```diff
@@ -109,54 +109,64 @@ public MetaStorageListener(KeyValueStorage storage) {
...
 
-                clo.result(!(cursorDesc == null) && cursorDesc.cursor().hasNext());
-            } else {
-                assert false : "Command was not found [cmd=" + command + ']';
+                    clo.result(!(cursorDesc == null) && cursorDesc.cursor().hasNext());
+                } else {
+                    assert false : "Command was not found [cmd=" + command + ']';
+                }
             }
+        } catch (NullPointerException e) {
```
Comment (sanpwc) [inline modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:167]: Same question, why we need this?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java [hunk 3]
```diff
@@ -444,16 +444,22 @@ void reset(final Status st) {
...
 
             if (this.done != null) {
-                this.done = (Status status) -> {
+                Closure newDone = (Status status) -> {
                     if (status.isOk()) {
                         node.getOptions().getRaftGrpEvtsLsnr().onNewPeersConfigurationApplied(resultPeers);
                     } else {
-                        node.getOptions().getRaftGrpEvtsLsnr().onReconfigurationError(status);
+                        node.getOptions().getRaftGrpEvtsLsnr().onReconfigurationError(status, () -> {
+                            // TODO: error handling for changePeersAsync https://ggsystems.atlassian.net/browse/IGN-19294
```
Comment (sanpwc) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/core/NodeImpl.java:452]: I'd rather expect such error handling to be on client side. There are few reasons for that, to mention some
- rebalance retry logic should be consistent with common sendWithRetry one. In case of inner handling (the one you've suggested above) it'll be possible to fail client side changePeersAsync with e.g. TimeoutException but still continue inner rebalance-retry logic. That seems to be invalid approach.
- It's incorrect to use metastore (for example in order to retrieve pending keys) from within NodeImpl.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java [hunk 4]
```diff
@@ -103,15 +136,75 @@ public void onNewPeersConfigurationApplied(List<PeerId> peers) {
...
-        // TODO: IGNITE-14873 implement this method
+    public void onReconfigurationError(Status status, Supplier<Void> rebalanceRunner) {
+        if (!busyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            if (status == null) {
+                // leader stepped down, so we are expecting RebalanceRaftGroupEventsListener.onLeaderElected to be called on a new leader.
+                LOG.info("Leader stepped down during the current rebalance for the partId = {}.");
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java:156]: What about partId param value?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java [hunk 5]
```diff
@@ -103,15 +136,75 @@ public void onNewPeersConfigurationApplied(List<PeerId> peers) {
...
+            LOG.warn("Error occurred during the current rebalance for partId = {}.", partId);
+
+            if (rebalanceAttempts.incrementAndGet() < REBALANCE_RETRY_THRESHOLD) {
+                rebalanceScheduler.schedule(() -> {
+                    LOG.info("Started {} attempt to retry the current rebalance for the partId = {}.", rebalanceAttempts.get(), partId);
+
+                    rebalanceRunner.get();
+                }, REBALANCE_RETRY_DELAY_MS, TimeUnit.MILLISECONDS);
+            } else {
+                LOG.info("Failed to perform the current rebalance for the partId = {}, canceling.", partId);
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java:172]: error

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java [hunk 6]
```diff
@@ -103,15 +136,75 @@ public void onNewPeersConfigurationApplied(List<PeerId> peers) {
...
+
+    }
+
+    /**
+     * This is a draft version of canceling a rebalance, this approach contains some serious flaws and must be revised.
+     */
+    private void cancelRebalance() {
+        Entry plannedEntry = metaStorageMgr.get(plannedPartAssignmentsKey(partId)).join();
+
+        if (plannedEntry.value() != null) {
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java:190]: We also should add an atomic logic of checking whether pending.revision is the same as the one we try to cancel, otherwise we might cancel new one.

Imaging following situation:
PENDING=[X]
Node A tries to rebalance data to X and fails all n retries -> it decides to cancel such X, thread sleep.
New Leander B elected that detects pending X and succeeds with it. On success it moves planned Y to pending.
Node A rebalance thread wakes up and cancels pending Y.

Comment (alievmirza) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java:190]: We discussed this approach with @kgusakov and we came to the conclusion, that is more preferable to throw some exception, when we haven't succeeded in the evaluation of the rebalance, instead of canceling it by removing `pending key`. We agreed that the current solution with removing pending key is not completed and may cause inconsistency in the current rebalance protocol. 

Also there is a case, when we can lost planned rebalance: 
1) Current leader retries failed rebalance
2) Current leader stops being leader for some reasons and sleeps
3) New leader performs rebalance and calls `RebalanceRaftGroupEventsListener#onNewPeersConfigurationApplied`
4) At this moment old leader wakes up and cancels the current rebalance, so it removes pending and writes to it planned key.
5) At this moment we receive `RebalanceRaftGroupEventsListener#onNewPeersConfigurationApplied` from the new leader, see that planned is empty, so we just delete pending key, but this is not correct to delete this key as far as the rebalance that is associated to the removed key hasn't been performed yet.

To sum up, I believe that on the current state of the rebalance feature, where we don't have proper failure handler, it is more convenient to throw an exception, so we can immediately react to  exceptional scenario on the TC, when we couldn't perform rebalance even after several retires.

@sanpwc WDYT?

Comment (alievmirza) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java:190]: We have discussed the solution in person and agreed to have infinite retry mechanism at this point of development, and to create separate ticket for rebalance canceling
